### ISSUE TITLE
Prewarming the cache with top CDN cached assets

### ISSUE THREAD

**Issue by jamesshannon** [WPO-Foundation/webpagetest#1096]
TLDR: WPT should provide an option to preload assets that are in the "global cache" (i.e., those assets that an average user is likely to have pre-cached before visiting a new domain) in order to allow for realistic first-visit measurements.

WPT doesn't accurately measure what a user will experience because it doesn't account for the "global cache" (a name I just made up). First impressions matter, and thus I typically avoid looking at the "second load" numbers, but unless the user is using Incognito mode then they're likely to have popular "global" CDN'ed libraries in their cache -- GPT, jQuery, AMP runtime, etc the first time they get around to visiting any given domain.

I ran into this the other day when it seemed to me that clickserve.dartsearch.net was unnecessarily doing an SSL hop (non-SSL to SSL) which wasted time on WPT scans. However, the PM pointed out to me that dartsearch.net sets HSTS and therefore practically all actual users will not experience the non-SSL redirect. (Since each WPT scan is fresh, then it doesn't remember HSTS headers.)

So, I've created a "[global cache warming page](https://rawgit.com/jamesshannon/845b6821a180d45c99ee10de88501fe2/raw/warm.html)" which WPT can load before it loads the page you're trying to measure performance of. The [source](https://gist.github.com/jamesshannon/845b6821a180d45c99ee10de88501fe2) is in gist and an example WPT script is below. In addition to dartsearch.net I used the webarchive to find the top requests with reasonable cache headers, and added them to the page.

A WPT scan using this approach is [here](https://www.webpagetest.org/result/171124_7V_228e7de6541bf06ba0843d7a000b9111/). Note that the page doesn't load the AMP runtime (unfortunately, it doesn't even show an AMP runtime request).

I'm posting this here to start a discussion of this approach, and to also request that similar functionality be built into WPT. I hope that there's a more efficient way to do this than load each asset before each request.

```
setActivityTimeout	5000
logData	0
navigate	https://rawgit.com/jamesshannon/845b6821a180d45c99ee10de88501fe2/raw/warm.html
logData	1 
navigate	https://landing.staging.lendingtree.com/lp/pla-amp.html
```

**Comment by pmeenan** [comment]
It's unlikely that something like this will ever be built directly into WPT.  The supported way to do it is with a script that loads a staging page with the resources you think are likely to be in a user's cache just like you have.

HSTS is probably cached better than resource caching but the hit rate for "common" CDN JS files is extremely low (particularly because of the number of versions available and the number of different CDN's).

I'll leave the issue open as informational for discussing different scripts people use (I have a hosted amp page for example that pre-warms the amp resources).

**Comment by andydavies** [comment]
The other thing that makes the hit rate for 'common' CDN JS files is Safari's double keying of the cache so even if two sites use the same resource from a third-party then they each 'end up with their own copy'
